Part of Issue246: Fees and rewards: Payout at end of next payout period

hewison-chris opened the pull request:
Paying the fees and rewards for the payout period preceding the one just ended.

Also still missing is:
1) dividing the validator rewards based on stake shares.
2) reducing validator rewards when missing signatures.
3) fixing some Ledger tests which do not yet include the rewards that have now been added.

hewison-chris committed:
Delay `Coinbase transaction` by one `payment period`

The same `Coinbase transaction` will be used to also pay the
`Validators` their block rewards. The rewards are reduced when not all
signatures are included in all blocks for the payment period, so time
must be given for Validators to share the block signatures.

hewison-chris committed:
Be more informative when `tx` is rejected

Make separate check and message for Coinbase or already in the tx pool.
The documentation was also updated to be correct.

Geod24 commented on doc/Database.md:
```diff
@@ -55,12 +55,13 @@ This table should be removed in favor of querying `validator` directly.
+| Field name      | SQL Type | D type     | Attributes  | Comment             |
+|-----------------|----------|------------|-------------|---------------------|
+| public_key      | TEXT     | PublicKey  | PRIMARY KEY |                     |
+| fee             | TEXT     | Enrollment |             | Should be `INTEGER` |
```
Is the D type correct ??

hewison-chris replied:
I'll add a commit to fix. It was wrong before.

hewison-chris replied:
Updated

Geod24 commented:
Typo in first commit msg (seperate).
Failure looks legit.

Geod24 commented on source/agora/consensus/Fee.d:
```diff
@@ -188,16 +192,18 @@ public class FeeManager
     ***************************************************************************/
 
-    private Amount[PublicKey] accumulated_fees;
+    private Amount[PublicKey][Height] accumulated_fees;
```
Don't think we ever clear this, do we ? Having a by height AA at least makes it easier to clean now.

hewison-chris replied:
The paid height is removed from the AA in `clearAccumulatedFees` 
```D
this.accumulated_fees.remove(height);
```

hewison-chris committed:
Correct SQL type of fee field in database

`Fee` is an `Amount` which has an underlying value type of `ulong`,
this can be stored as an INTEGER in the `SQLite3` database.

hewison-chris commented:
Fixed typo and added commit to change the fee SQL type to INTEGER.
Still need to fix the failing tests though.

linked0 commented on source/agora/common/Amount.d:
```diff
@@ -39,7 +39,7 @@ public struct Amount
     public static immutable Amount UnitPerCoin   = Amount(10_000_000, true);
     /// Maximum number of BOA coins that can ever be in circulation
-    public static immutable ulong  MaxCoinSupply = 500_000_000;
+    public static immutable ulong  MaxCoinSupply = 5_000_000_000;
```
Wouldn't it be 4_950_000_000?

hewison-chris replied:
Yes you are right. Updated.

hewison-chris replied:
In fact this is from the other PR which I rebased this one on. Will update https://github.com/bosagora/agora/pull/2415

hewison-chris replied:
https://github.com/bosagora/agora/pull/2415/commits/8851558aea35bb92168f88d03b6b33dfea1bd85c

hewison-chris committed:
Store the Validator stake in the database

For convenience when the amount that was staked for a validator is
needed.

hewison-chris commented:
> Typo in first commit msg (seperate).
> Failure looks legit.

Fixed typo and the test that was failing.

hewison-chris committed:
wip - storing fees per height to enable calculating distribution to validators
per block signed.

hewison-chris commented:
Closing this as other PRs are more up to date with incremental changes.
